python/ql/src/semmle/python/types/Exceptions.qll

```diff
@@ -260,40 +260,69 @@ class ExceptFlowNode extends ControlFlowNode {
         )
     }
 
-    private predicate handledObject(Object obj, ClassObject cls, ControlFlowNode origin) {
+    private predicate handledObject_objectapi(Object obj, ClassObject cls, ControlFlowNode origin) {
         this.getType().refersTo(obj, cls, origin)
         or
         exists(Object tup |
-            this.handledObject(tup, theTupleType(), _) |
-            element_from_tuple(tup).refersTo(obj, cls, origin)
+            this.handledObject_objectapi(tup, theTupleType(), _) |
+            element_from_tuple_objectapi(tup).refersTo(obj, cls, origin)
+        )
+    }
+    
+    private predicate handledObject(Value val, ClassValue cls, ControlFlowNode origin) {
+        val.getClass() = cls and
+        (
+            this.getType().pointsTo(val, origin)
+            or
+            exists(Value tup |
+                this.handledObject(val, ClassValue::tuple(), _) |
+                element_from_tuple(tup).pointsTo(val, origin)
+            )
         )
     }
 
     /** Gets the inferred type(s) that are handled by this node, splitting tuples if possible. */
     pragma [noinline]
-    predicate handledException(Object obj, ClassObject cls, ControlFlowNode origin) {
-        this.handledObject(obj, cls, origin) and not cls = theTupleType()
+    predicate handledException_objectapi(Object obj, ClassObject cls, ControlFlowNode origin) {
+        this.handledObject_objectapi(obj, cls, origin) and not cls = theTupleType()
         or
         not exists(this.getNode().(ExceptStmt).getType()) and obj = theBaseExceptionType() and cls = theTypeType() and
         origin = this
     }
+    
+    /** Gets the inferred type(s) that are handled by this node, splitting tuples if possible. */
+    pragma [noinline]
+    predicate handledException(Value val, ClassValue cls, ControlFlowNode origin) {
+        this.handledObject(val, cls, origin) and not cls = ClassValue::tuple()
+        or
+        not exists(this.getNode().(ExceptStmt).getType()) and val = ClassValue::baseException() and cls = ClassValue::type() and
+        origin = this
+    }
+    
+    
 
     /** Whether this `except` handles `cls` */
     predicate handles(ClassObject cls) {
         exists(ClassObject handled |
-            this.handledException(handled, _, _) |
+            this.handledException_objectapi(handled, _, _) |
             cls.getAnImproperSuperType() = handled
         )
     }
 
 }
 
-private ControlFlowNode element_from_tuple(Object tuple) {
+private ControlFlowNode element_from_tuple_objectapi(Object tuple) {
     exists(Tuple t |
         t = tuple.getOrigin() and result = t.getAnElt().getAFlowNode()
     )
 }
 
+private ControlFlowNode element_from_tuple(Value tuple) {
```

Comment:
I wonder if it would suffice to make the argument a `TupleValue`, and use `getAnElement()` and `getOrigin` to get the same behaviour. (And alter line 279 accordingly).

Reply:
Looking at this again, I'm not sure I really .. like this predicate? `element_from_tuple` takes a tuple *value* and picks out a CFG that corresponds to the value's AST source's elements. Which is just *weird*.

Reply:
Right. This is essentially what `getItem` does on the `TupleValue` class (this is what I thought was called `getAnElement`).

So what I'm suggesting is to use `getItem(_)` on the `TupleValue` directly, instead of on the corresponding control-flow node.

You'll also need a way of extracting the origin of the tuple element. I think it should be possible to extract this using `pointsTo`. If not, we'll have to add it as a method on relevant `Values` (e.g. something like `True` won't have an origin, since it's just a constant literal, but a specific tuple will. Essentially, this boils down to whether the underlying IPA type points back to a control-flow node).

Reply:
I think that's the right approach, yeah. I tried this in a different PR and it seems to work fine. I'll swap it out now.

Reply:
Should be solved now. :)